src/consensus/tower.zig

```diff
@@ -0,0 +1,3177 @@
+                    // newer snapshots on validator restart
+                    self.logger.warn().logf(
+                        \\bank_forks doesn't have corresponding data for the stray restored last vote({}),
+                        \\meaning some inconsistency between saved tower and ledger.
+                    , .{last_voted_slot});
+                }
+                break :blk SortedSet(u64).init(allocator);
+            } else return error.NoAncestorsFoundForLastVote;
+        };
+
+        const switch_slot_ancestors = ancestors.get(switch_slot) orelse
+            return error.NoAncestorsFoundForSwitchSlot;
+
+        if (switch_slot == last_voted_slot or switch_slot_ancestors.contains(last_voted_slot)) {
+            // If the `switch_slot is a descendant of the last vote,
+            // no switching proof is necessary
+            return SwitchForkDecision{ .same_fork = {} };
+        }
+
+        if (last_vote_ancestors.contains(switch_slot)) {
+            if (self.isStrayLastVote()) {
+                // This peculiar corner handling is needed mainly for a tower which is newer than
+                // blockstore. (Yeah, we tolerate it for ease of maintaining validator by operators)
+                // This condition could be introduced by manual ledger mishandling,
+                // validator SEGV, OS/HW crash, or plain No Free Space FS error.
+
+                // When we're in this clause, it basically means validator is badly running
+                // with a future tower while replaying past slots, especially problematic is
+                // last_voted_slot.
+                // So, don't re-vote on it by returning pseudo FailedSwitchThreshold, otherwise
+                // there would be slashing because of double vote on one of last_vote_ancestors.
+                // (Well, needless to say, re-creating the duplicate block must be handled properly
+                // at the banking stage: https://github.com/solana-labs/solana/issues/8232)
+                //
+                // To be specific, the replay stage is tricked into a false perception where
+                // last_vote_ancestors is AVAILABLE for descendant-of-`switch_slot`,  stale, and
+                // stray slots (which should always be empty_ancestors).
+                //
+                // This is covered by test_future_tower_* in local_cluster
+                return SwitchForkDecision{ .failed_switch_threshold = .{ 0, total_stake } };
+            } else return error.NoAncestorsFoundForLastVote;
+        }
+
+        // By this point, we know the `switch_slot` is on a different fork
+        // (is neither an ancestor nor descendant of `last_vote`), so a
+        // switching proof is necessary
+        const switch_proof = Hash.ZEROES;
+        var locked_out_stake: u64 = 0;
+        var locked_out_vote_accounts = SortedSet(Pubkey).init(allocator);
+        var iterator = descendants.iterator();
+        while (iterator.next()) |descendant| {
```

Comment:
Unless you have a particular reason against it, I would suggest making the descendants map an array hash map if you're going to iterate over it at all.

Reply:
[4be8a58](https://github.com/Syndica/sig/pull/664/commits/4be8a583187a899c3bd4374e50ce9ffe7e142662)